### ISSUE TITLE
Misleading "error: source trait is private" when a trait is not imported

### ISSUE THREAD

**Issue by gkoz** [rust-lang/rust#22050]
This example uses `rust-crypto 0.2.14`. Failing to import the `Encryptor` trait that implements `encrypt` causes an unexpected error message. The trait isn't private and importing it compiles as expected.

``` rust
extern crate crypto;

use crypto::{ buffer, aes, blockmodes };
use crypto::buffer::{ ReadBuffer, WriteBuffer };

fn enc(key: &[u8], iv: &[u8], data: &[u8]) {
    // Don't import the trait that implements encrypt()
    //use crypto::symmetriccipher::Encryptor;

    let mut encryptor = aes::cbc_encryptor(aes::KeySize::KeySize256,
                key, iv, blockmodes::PkcsPadding);

    let mut read_buffer = buffer::RefReadBuffer::new(data);
    let mut buffer = [0; 4096];
    let mut write_buffer = buffer::RefWriteBuffer::new(&mut buffer);

    // Expected error: type `_` does not implement any method in scope named `encrypt`
    // Get error: source trait is private
    encryptor.encrypt(&mut read_buffer, &mut write_buffer, true);
}

fn main() {
}
```

```
src/main.rs:19:5: 19:65 error: source trait is private
src/main.rs:19     encryptor.encrypt(&mut read_buffer, &mut write_buffer, true);
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

`rustc 1.0.0-nightly (706be5ba1 2015-02-05 23:14:28 +0000)`

**Comment by GuillaumeGomez** [comment]
The new error message would need to be something like "error: source trait not found" I guess ?

**Comment by shepmaster** [comment]
[Another person got bit](http://stackoverflow.com/q/30448596/155423)

**Comment by Wilfred** [comment]
For what it's worth, I was confused by this issue today.

**Comment by steveklabnik** [comment]
me too

**Comment by GuillaumeGomez** [comment]
So should I replace it by "error: source trait not found" ?

**Comment by steveklabnik** [comment]
I would argue it should be 

    error: use of undeclared trait name `Foo`

Because that's what happens when you do something like

```
fn foo<T: Foo>() {}

fn main() {

}
```

**Comment by Wilfred** [comment]
@GuillaumeGomez That would be clearer, yep. It could also suggest adding an `extern crate...` or an entry in Cargo.toml, although perhaps this occurs in other cases that haven't occurred to me.

**Comment by steveklabnik** [comment]
well, i guess that's different code than the OP's. Basically, the error should mirror what happens when it doesn't exist

**Comment by gkoz** [comment]
Am I wrong to expect
```
error: no method named `encrypt` found for type `` in the current scope
```

Possibly with a hint about importing the correct trait. Y'know like it usually happens.

**Comment by jethrogb** [comment]
Also expecting the error message @gkoz mentions. What is different about this than, for example:
```rust
let mut buf=[0u8;32];
(&mut buf[..]).write_all(&b"test");
```
which yields
```
<anon>:3:20: 3:39 error: no method named `write_all` found for type `&mut [u8]` in the current scope
<anon>:3     (&mut buf[..]).write_all(&b"test");
                            ^~~~~~~~~~~~~~~~~~~
<anon>:3:20: 3:39 help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
<anon>:3:20: 3:39 help: candidate #1: use `std::io::Write`
```

**Comment by naufraghi** [comment]
Got the same problem (1.8.0):

```rust
//extern crate num;
extern crate nalgebra;

let (sin_pitch, cos_pitch) = pitch.sin_cos();
```

Gives the user no hints where to search for the `sin_cos` impl.

If the compiler knows which is the private trait that is missing, an useful message could show that name as @steveklabnik suggested. Even better with the suggested `extern crate`.

**Comment by petrochenkov** [comment]
This should be fixed in https://github.com/rust-lang/rust/pull/32073 & https://github.com/rust-lang/rust/pull/31920

**Comment by jseyfried** [comment]
This is a duplicate of #21670 -- the example here is covered by the test from #32073.
